### DISCUSSION TITLE
Shared notes w/o javascript

### DISCUSSION THREAD

**Discussion by sigaloid** [TriliumNext/Trilium#2467] [General]
![Screenshot from 2021-12-24 11-25-22](https://user-images.githubusercontent.com/69441971/147364506-8f4a0bf2-7259-4b1e-a1a4-acb16da039e5.png)

I was able to add mermaid diagrams to shared notes pretty quickly but it does do the rendering client-side, meaning there is javascript. I did notice a lack of javascript in shared notes and I'm actually pretty glad about that, so does that mean math+mermaid diagrams+others have to be rendered serverside?  I think that can be done with JSDom but I also am pretty new to this language.

**Comment by sigaloid** [comment]
Okay, so mermaid diagrams cannot be rendered serverside as per https://github.com/mermaid-js/mermaid/issues/559 and https://github.com/mermaid-js/mermaid/issues/1183 as rendering relies on clientside DOM access. [Mathjax does have a Node API](https://github.com/mathjax/MathJax-node) so I think you can simply loop through the items in the dom via `document.querySelectorAll("span")`, if the class name is math-tex, render as SVG (or HTML) and replace that node's inner HTML with it.

**Comment by zadam** [comment]
Shared notes are rendered on the server side so that it's easily processable by e.g. search engines.

But I'm not sure if this is useful/needed in case of Math/Mermaid. I think I would probably do it first in frontend JavaScript and only if there are significant downsides it can be moved into backend.

**Reply by sigaloid** [reply]
Ah, okay, shall I make PR for mermaid diagrams to be added?

And if so, is there a mermaid.js hosted within trilium I can link the script to instead of just the CDN? (maybe it needs to be added to the share route?0

**Reply by zadam** [reply]
Yes, PR would be appreciated :)

there's a library/mermaid.min.js and library/mermaid.min.css which you can use.

**Reply by sigaloid** [reply]
Done! #2471
